File: weltwerk/verify/orchestrator.py
# SPDX-License-Identifier: AGPL-3.0-only
"""
orchestrator.py — a thin Epistemic Runtime Orchestrator: make the verified tools interchangeable behind two
chokepoints, adding NO new authority.

The tools (`residual_channel`, `claim_ledger`, `certificate_compiler`, `frontier_gate`, the engines) already
exist and are tested. This layer does one narrow thing: it standardizes how they are *called* and *composed*,
through exactly two type gates that the rest of the repo already enforces piecewise —

  • ANSWERS go through `AnalysisResult` — every registered tool's `analyze()` returns the honesty contract
    (scope + ≥1 limitation), checked again at the orchestration boundary. A tool that emits a bare verdict is
    rejected here.
  • ACTIONS go through `Grounded[T]` — `enact()` runs a side effect only if the value is `Grounded` by a
    verifier-issued proof; an ungrounded action raises `UngroundedError` before any effect.

What it deliberately is NOT: it is not a "platform" that makes the tools more capable, and it does not produce a
single confidence scalar. `panel()` returns *many witnesses on one request, side by side* — it never averages or
ranks them into one number (the project's standing refusal). The orchestrator is a *router + two chokepoints*; it
grants no truth the underlying tool did not. `orchestration ≠ authority`; `router ≠ verifier`;
`composition ≠ capability`; `integrity ≠ truth`.
"""
from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Protocol, Tuple, runtime_checkable

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from artifacts import AnalysisResult, Finding, Limitation                      # noqa: E402
from epistemic_types import Grounded, Grounding, UngroundedError               # noqa: E402


class UnknownToolError(KeyError):
    """Requested a tool that is not registered — a typed error, never a silent default."""


@runtime_checkable
class EpistemicTool(Protocol):
    name: str
    def analyze(self, request: dict) -> AnalysisResult: ...


# ---- adapters: wrap each existing tool to the uniform (request → AnalysisResult) interface --------
@dataclass(frozen=True)
class ResidualTool:
    """request: {'samples': [(x,y,z), …], optional 'misspec_fns'}. Wraps residual_channel.audit."""
    name: str = "residual_channel"
    def analyze(self, request: dict) -> AnalysisResult:
        from residual_channel import audit
        r = audit(request["samples"], misspec_fns=request.get("misspec_fns"))
        return r.as_analysis()


@dataclass(frozen=True)
class FrontierTool:
    """request: {'m_novel': float, 'ci': (lo,hi)}. Wraps frontier_gate.FrontierGate."""
    name: str = "frontier_gate"
    def analyze(self, request: dict) -> AnalysisResult:
        from frontier_gate import FrontierGate
        g = FrontierGate()
        return g.as_analysis(g.decide(request["m_novel"], request["ci"]))


@dataclass(frozen=True)
class LedgerTool:
    """request: {'ledger': (Claim, …)}. Wraps claim_ledger.audit_ledger into the honesty contract."""
    name: str = "claim_ledger"
    def analyze(self, request: dict) -> AnalysisResult:
        from claim_ledger import audit_ledger
        a = audit_ledger(request["ledger"])
        findings = (
            Finding("LEDGER_HONEST", "claim-ledger", f"honest={a['honest']} counts={a['counts']}"),
            Finding("VIOLATIONS", "claim-ledger",
                    f"off_ladder={a['off_ladder']} missing={a['missing_falsifier_or_boundary']}"),
        )
        limitations = (
            Limitation("claim-ledger", "the audit checks FORM (graded + falsifiable + bounded), not the truth "
                                       "of the claims; integrity ≠ truth"),
        )
        return AnalysisResult(source_trace=(), scope="claim-ledger", findings=findings, limitations=limitations)


@dataclass(frozen=True)
class CertificateTool:
    """request: {'init','successors','universe','cert'}. Wraps certificate_compiler.check_certificate."""
    name: str = "certificate_compiler"
    def analyze(self, request: dict) -> AnalysisResult:
        from certificate_compiler import check_certificate, minimal_reason
        r = check_certificate(request["init"], request["successors"], request["universe"], request["cert"])
        findings = (
            Finding("INDUCTIVE", "inductive-certificate", f"valid={r.valid} ({minimal_reason(r)})"),
            Finding("WITNESS", "inductive-certificate", f"{r.witness!r}" if not r.valid else "none"),
        )
        limitations = (
            Limitation("inductive-certificate", "checked over a supplied finite universe (pure-stdlib); the "
                                                "size-independent symbolic check is the z3 path. checking ≠ finding"),
            Limitation("inductive-certificate", "the checker verifies a GIVEN invariant; it does not derive one"),
        )
        return AnalysisResult(source_trace=(), scope="inductive-certificate", findings=findings, limitations=limitations)


# ---- the orchestrator: registry + the two chokepoints --------------------------------------------
@dataclass
class EpistemicRuntimeOrchestrator:
    tools: Dict[str, EpistemicTool] = None

    def __post_init__(self):
        if self.tools is None:
            self.tools = {}

    def register(self, tool: EpistemicTool) -> "EpistemicRuntimeOrchestrator":
        if not isinstance(tool, EpistemicTool):
            raise TypeError(f"{tool!r} does not satisfy the EpistemicTool protocol (name + analyze)")
        self.tools[tool.name] = tool
        return self

    def analyze(self, tool_name: str, request: dict) -> AnalysisResult:
        """ANSWER chokepoint: dispatch, and enforce the honesty contract at the boundary."""
        if tool_name not in self.tools:
            raise UnknownToolError(f"no tool '{tool_name}' registered ({sorted(self.tools)})")
        r = self.tools[tool_name].analyze(request)
        if not isinstance(r, AnalysisResult):
            raise TypeError(f"tool '{tool_name}' did not return an AnalysisResult")
        if not r.scope or len(r.limitations) < 1:
            raise ValueError(f"tool '{tool_name}' emitted a result without the honesty contract")
        return r

    def panel(self, calls: List[Tuple[str, dict]]) -> Dict[str, AnalysisResult]:
        """Many witnesses on one question, side by side — NO aggregation, NO scalar, no global winner."""
        return {name: self.analyze(name, req) for name, req in calls}

    def enact(self, value: Any, proof: Grounding, action: Callable[[Any], Any]) -> Any:
        """ACTION chokepoint: run `action` only on a value GROUNDED by a verifier-issued proof; an ungrounded
        value raises UngroundedError before any effect. The orchestrator grants no authority — the proof does."""
        g = Grounded.ground(value, proof)        # raises UngroundedError if not grounded
        return action(g.value)


def default_orchestrator() -> EpistemicRuntimeOrchestrator:
    o = EpistemicRuntimeOrchestrator()
    for t in (ResidualTool(), FrontierTool(), LedgerTool(), CertificateTool()):
        o.register(t)
    return o


def main():
    print("orchestrator.py — Epistemic Runtime Orchestrator (router + two chokepoints, no new authority)\n")
    from residual_channel import demo_gen_channel
    from claim_ledger import Claim
    from certificate_compiler import ConstraintCertificate, chain
    from epistemic_types import Attested

    o = default_orchestrator()
    init, succ, uni = chain(6)
    calls = [
        ("residual_channel", {"samples": demo_gen_channel()}),
        ("frontier_gate", {"m_novel": 0.5, "ci": (0.42, 0.58)}),
        ("claim_ledger", {"ledger": (Claim("D1", "E reproduces", "ESTABLISHED", "M", "the magnitude", "a failed rep"),)}),
        ("certificate_compiler", {"init": init, "successors": succ, "universe": uni,
                                  "cert": ConstraintCertificate("ok", lambda i: 0 <= i <= 6, lambda i: True)}),
    ]
    for name, res in o.panel(calls).items():
        print(f"  [{name}] scope={res.scope!r}  findings={len(res.findings)}  limitations={len(res.limitations)}")
        print(f"       → {res.findings[0].detail if res.findings else '(none)'}")

    log: List = []
    o.enact("apply", Attested(True, "verifier-grounded"), lambda v: log.append(v))
    try:
        o.enact("apply", Attested(False, "ungrounded"), lambda v: log.append(v))
    except UngroundedError:
        pass
    print(f"\n  enact: committed {log} (ungrounded action refused before effect)")
    print("  every answer is an AnalysisResult; every action passes the Grounded gate; the panel stays plural.")
    print("  orchestration ≠ authority; router ≠ verifier; composition ≠ capability; integrity ≠ truth.")


if __name__ == "__main__":
    main()

File: Rust/src/orchestrator.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! The Epistemic Runtime Orchestrator — a router with two chokepoints, and NO new authority.
//!
//! * **Answer chokepoint.** Every tool returns an [`AnalysisResult`], which is honest *by construction*
//!   (see [`crate::artifacts`]). The orchestrator cannot emit, and a tool cannot build, a dishonest answer.
//!   [`Orchestrator::panel`] reports many witnesses side by side — there is NO fused confidence scalar.
//! * **Action chokepoint.** [`Orchestrator::enact`] runs an action only on a value a verifier-issued
//!   [`Grounding`] grounds; otherwise it refuses before any effect.
//!
//! `router ≠ verifier`; `composition ≠ capability`; `integrity ≠ truth`.

use std::collections::BTreeMap;
use std::error::Error;
use std::fmt;

use crate::artifacts::{AnalysisResult, Finding, Limitation};
use crate::claim_ledger::{audit_ledger, Claim};
use crate::epistemic_types::{enact, Grounding, UngroundedError};
use crate::frontier_gate::FrontierGate;
use crate::residual_channel::{audit, Sample};
use crate::coupling_audit::{audit_coupling, CouplingInput};

/// The request a tool consumes. One variant per fundamental tool.
pub enum Request {
    Residual { samples: Vec<Sample>, reps: usize, seed: u64 },
    Frontier { m_novel: f64, ci: (f64, f64) },
    Ledger { claims: Vec<Claim> },
    Coupling(Box<CouplingInput>),
}

/// A registered tool: a name + a total function into the honesty contract.
pub trait EpistemicTool {
    fn name(&self) -> &str;
    fn analyze(&self, req: &Request) -> AnalysisResult;
}

fn unsupported(tool: &str) -> AnalysisResult {
    AnalysisResult::new(
        "error",
        vec![Finding::new("UNSUPPORTED_REQUEST", "error", tool.to_string())],
        vec![Limitation::new("error", "this tool received a request variant it does not handle")],
    )
    .expect("unsupported() always satisfies the honesty contract")
}

/// Confounder-conditioned dependence audit.
pub struct ResidualTool;
impl EpistemicTool for ResidualTool {
    fn name(&self) -> &str {
        "residual_channel"
    }
    fn analyze(&self, req: &Request) -> AnalysisResult {
        match req {
            Request::Residual { samples, reps, seed } => {
                audit(samples, *reps, *seed, 4.0, 0.005, &[]).as_analysis()
            }
            _ => unsupported(self.name()),
        }
    }
}

/// Bounded metric-deflation gate.
pub struct FrontierTool;
impl EpistemicTool for FrontierTool {
    fn name(&self) -> &str {
        "frontier_gate"
    }
    fn analyze(&self, req: &Request) -> AnalysisResult {
        match req {
            Request::Frontier { m_novel, ci } => {
                let d = FrontierGate::default().decide(*m_novel, *ci);
                AnalysisResult::new(
                    "frontier",
                    vec![
                        Finding::new("REGIME", "frontier", format!("{:?}", d.regime)),
                        Finding::new("ACTION", "frontier", d.action.as_str()),
                        Finding::new("M_NOVEL", "frontier", format!("{:.4}", d.m_novel)),
                    ],
                    vec![Limitation::new(
                        "frontier",
                        "m_novel is an estimate; pivot != guaranteed-escape",
                    )],
                )
                .expect("frontier analysis satisfies the honesty contract")
            }
            _ => unsupported(self.name()),
        }
    }
}

/// Graded claim-ledger audit.
pub struct LedgerTool;
impl EpistemicTool for LedgerTool {
    fn name(&self) -> &str {
        "claim_ledger"
    }
    fn analyze(&self, req: &Request) -> AnalysisResult {
        match req {
            Request::Ledger { claims } => {
                let a = audit_ledger(claims);
                AnalysisResult::new(
                    "ledger",
                    vec![
                        Finding::new("HONEST", "ledger", a.honest.to_string()),
                        Finding::new("COUNTS", "ledger", format!("{:?}", a.counts)),
                        Finding::new("MISSING_BOUNDARY", "ledger", format!("{:?}", a.missing_boundary)),
                    ],
                    vec![Limitation::new("ledger", "honest = graded + falsifiable; honest != true")],
                )
                .expect("ledger analysis satisfies the honesty contract")
            }
            _ => unsupported(self.name()),
        }
    }
}

/// Forbidden-coupling firewall — the taxonomy (AIR_GAP_HELD / OBSERVER_CONTAMINATION / CONFOUNDED_ARTIFACT /
/// UNIDENTIFIABLE) layered on the residual-channel core.
pub struct CouplingTool;
impl EpistemicTool for CouplingTool {
    fn name(&self) -> &str {
        "coupling_audit"
    }
    fn analyze(&self, req: &Request) -> AnalysisResult {
        match req {
            Request::Coupling(inp) => audit_coupling(inp).as_analysis(),
            _ => unsupported(self.name()),
        }
    }
}

/// Tool dispatch failure.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OrchestratorError {
    UnknownTool(String),
}

impl fmt::Display for OrchestratorError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            OrchestratorError::UnknownTool(n) => write!(f, "no tool '{}' registered", n),
        }
    }
}

impl Error for OrchestratorError {}

/// A router over registered [`EpistemicTool`]s. Adds no authority — it dispatches answers and gates actions.
pub struct Orchestrator {
    tools: BTreeMap<String, Box<dyn EpistemicTool>>,
}

impl Orchestrator {
    pub fn new() -> Self {
        Self { tools: BTreeMap::new() }
    }

    pub fn register(&mut self, tool: Box<dyn EpistemicTool>) -> &mut Self {
        self.tools.insert(tool.name().to_string(), tool);
        self
    }

    /// ANSWER chokepoint: dispatch by name. The returned `AnalysisResult` is honest by construction.
    pub fn analyze(&self, name: &str, req: &Request) -> Result<AnalysisResult, OrchestratorError> {
        match self.tools.get(name) {
            Some(t) => Ok(t.analyze(req)),
            None => Err(OrchestratorError::UnknownTool(name.to_string())),
        }
    }

    /// Many witnesses on one question, side by side — NO aggregation, NO scalar, no global winner.
    pub fn panel(
        &self,
        calls: &[(String, Request)],
    ) -> Vec<(String, Result<AnalysisResult, OrchestratorError>)> {
        calls.iter().map(|(n, r)| (n.clone(), self.analyze(n, r))).collect()
    }

    /// ACTION chokepoint: run `action` only on a value the `proof` grounds; else `Err` before any effect.
    pub fn enact<T, P, R, F>(&self, value: T, proof: &P, action: F) -> Result<R, UngroundedError>
    where
        P: Grounding,
        F: FnOnce(T) -> R,
    {
        enact(value, proof, action)
    }

    pub fn tool_names(&self) -> Vec<String> {
        self.tools.keys().cloned().collect()
    }
}

impl Default for Orchestrator {
    fn default() -> Self {
        Self::new()
    }
}

/// An orchestrator with the fundamental tools registered.
pub fn default_orchestrator() -> Orchestrator {
    let mut o = Orchestrator::new();
    o.register(Box::new(ResidualTool))
        .register(Box::new(FrontierTool))
        .register(Box::new(LedgerTool))
        .register(Box::new(CouplingTool));
    o
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::claim_ledger::Grade;
    use crate::epistemic_types::Attested;
    use crate::residual_channel::demo_gen_channel;

    #[test]
    fn unknown_tool_errors() {
        let o = default_orchestrator();
        let r = o.analyze("nope", &Request::Frontier { m_novel: 1.0, ci: (0.9, 1.1) });
        assert!(matches!(r, Err(OrchestratorError::UnknownTool(_))));
    }

    #[test]
    fn answers_are_honest_by_construction() {
        let o = default_orchestrator();
        let a = o
            .analyze("residual_channel", &Request::Residual { samples: demo_gen_channel(2000, 3, 2), reps: 40, seed: 0 })
            .unwrap();
        assert!(!a.scope().is_empty());
        assert!(!a.limitations().is_empty());
    }

    #[test]
    fn panel_carries_no_scalar() {
        let o = default_orchestrator();
        let calls = vec![
            ("frontier_gate".to_string(), Request::Frontier { m_novel: 0.5, ci: (0.3, 0.8) }),
            (
                "claim_ledger".to_string(),
                Request::Ledger {
                    claims: vec![Claim::new("D1", "E reproduces", Grade::Established, "m", "the magnitude", "a failed rep")],
                },
            ),
        ];
        let panel = o.panel(&calls);
        // one witness per call; the panel is a Vec of named results — no fused score anywhere
        assert_eq!(panel.len(), 2);
        assert!(panel.iter().all(|(_n, r)| r.is_ok()));
    }

    #[test]
    fn action_chokepoint_gates() {
        let o = default_orchestrator();
        assert!(o.enact("apply", &Attested::new(true, "grounded"), |v| v).is_ok());
        assert!(o.enact("apply", &Attested::new(false, "ungrounded"), |v| v).is_err());
    }
}
